packages/amplify-graphql-model-transformer/src/resources/amplify-dynamodb-table/amplify-dynamo-model-resource-generator.ts

```diff
@@ -58,52 +68,82 @@ export class AmplifyDynamoModelResourceGenerator extends DynamoModelResourceGene
             'dynamodb:UpdateTimeToLive',
           ],
           resources: [
+            // eslint-disable-next-line no-template-curly-in-string
             cdk.Fn.sub('arn:aws:dynamodb:${AWS::Region}:${AWS::AccountId}:table/*-${apiId}-${envName}', {
-              apiId: ctx.api.apiId,
-              envName: ctx.synthParameters.amplifyEnvironmentName,
+              apiId: context.api.apiId,
+              envName: context.synthParameters.amplifyEnvironmentName,
             }),
           ],
         }),
-      );
-    }
-
-    this.generateResolvers(ctx);
-  }
-
-  protected createCustomProviderResource(scope: Construct, context: TransformerContextProvider): void {
-    // Policy that grants access to Create/Update/Delete DynamoDB tables
-    this.ddbManagerPolicy = new aws_iam.Policy(scope, 'CreateUpdateDeleteTablesPolicy');
-
-    const lambdaCode = aws_lambda.Code.fromAsset(
-      path.join(__dirname, '..', '..', '..', 'lib', 'resources', 'amplify-dynamodb-table', 'amplify-table-manager-lambda'),
-    );
-
-    // lambda that will handle DDB CFN events
-    const gsiOnEventHandler = new aws_lambda.Function(scope, ResourceConstants.RESOURCES.TableManagerOnEventHandlerLogicalID, {
-      runtime: aws_lambda.Runtime.NODEJS_18_X,
-      code: lambdaCode,
-      handler: 'amplify-table-manager-handler.onEvent',
-      timeout: Duration.minutes(14),
+      ],
     });
 
-    // lambda that will poll for provisioning to complete
-    const gsiIsCompleteHandler = new aws_lambda.Function(scope, ResourceConstants.RESOURCES.TableManagerIsCompleteHandlerLogicalID, {
-      runtime: aws_lambda.Runtime.NODEJS_18_X,
-      code: lambdaCode,
-      handler: 'amplify-table-manager-handler.isComplete',
-      timeout: Duration.minutes(14),
+    // Note: The isCompleteRole and onEventRole are similar enough that you might ask "why not just use a single role?"
+    // 1. Doing so creates a circular dependency between someCombinedRole <-> waiterStateMachine
+    // 2. The isCompleteHandler doesn't need permissions to invoke the waiterStateMachine.
+
+    // Role assumed by the isCompleteHandler.
+    // We want to avoid the auto-generated default policy for this to avoid unnecessary deployment time
+    // slowdowns, hence the `withPolicyUpdates()`
+    const isCompleteRole = new aws_iam.Role(scope, 'AmplifyManagedTableIsCompleteRole', {
+      assumedBy: new aws_iam.ServicePrincipal('lambda.amazonaws.com'),
+      managedPolicies: [aws_iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole')],
+      inlinePolicies: {
+        CreateUpdateDeleteTablesPolicy: lambdaPolicyDocument,
+      },
+    }).withoutPolicyUpdates();
+
+    // Role assumed by the onEventHandler (custom resource entry point).
+    // We need to keep this open to modification so that waiter state machine can grant it
+    // invocation permissions below, hence no `withoutPolicyUpdates()`
```

Comment:
For future PR: Could we gain initial deployment time improvements with an inline policy that added the necessary invocation permissions, rather than keeping it open? Or is there not a duplicate role created for this particular lambda?

Reply:
There's only one role created for the `onEventHandler`. But in addition to the inline policy defined, there is a standalone IAM Policy created by the CDK through `waiterStateMachine.grantStartExecution(onEventHandler);` below.

That policy is:
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": "states:StartExecution",
            "Resource": "<state-machine-arn>",
            "Effect": "Allow"
        }
    ]
}
```

To answer your question, yes it would be possible for us to inline this policy into the role, reducing the need for the standalone policy. The theoretical upper bound of deployment speed improvement would be 15-20 seconds. Practically, it would likely shave off only a few seconds due to resources being provisioned in parallel.

I didn't include this because doing so led to circular dependency issues in testing. But from a fresh look, I'm not seeing how that would happen, so I may have been holding it wrong. Depending on the scope of change from the remaining comments, I may try to sneak that in. If not, it's definitely worth a follow up.

Reply:
Likely possible without a circular dependency, but it's going to require more changes than I'd like to include in this PR. Will follow up with it.

Reply:
incremental improvement seems like a good approach here.